### PR TITLE
Make TNT faster by using `VoxelManip()` for removing nodes.

### PR DESCRIPTION
**Pull request by red-001**
For testing I changed `tnt_radius` in minetest.conf to 20. Without my patch an explosion took 6 seconds and with it it took one second.

### CODE REVIEW

**Comment by paramat** [comment]
If people complain you might also want to queue 'on destruct' and 'after destruct', although 'on destruct' is usually called before removing node so may have problems.

File: mods/tnt/init.lua [hunk 1]
```diff
@@ -201,13 +203,22 @@ local function explode(pos, radius)
...
 		vi = vi + 1
 	end
 	end
 	end
+	
+	vm:set_data(data)
+	vm:update_liquids()
+	vm:write_to_map()
+	vm:update_map()
+	for i, data in ipairs(on_blast_queue) do
```
Comment (kilbith) [inline mods/tnt/init.lua]: Where is `i` used ? Please ignore it with an underscore.

Comment (red-001) [inline mods/tnt/init.lua]: done

**Comment by red-001** [comment]
@kilbith done.

**Comment by paramat** [comment]
Needs input on whether 'on/after destruct' functions should be queued also.

**Comment by red-001** [comment]
It should be rather easy for mods to use on_blast instead. It could break so mods that are no longer maintained .

**Comment by C1ffisme** [comment]
This should also help with server-freezing chain reactions. :+1:

**Comment by paramat** [comment]
The possibility of destroying locked chests or protected nodes shoud be considered too.

**Comment by tenplus1** [comment]
Add something like this to check node and if it's a chest drop all items inside.
```
-- if chest then drop items inside
if n == "default:chest"
or n == "3dchest:chest"
or n == "bones:bones" then

	local meta = minetest.get_meta(p)
	local inv  = meta:get_inventory()

	for i = 1, inv:get_size("main") do

		local m_stack = inv:get_stack("main", i)
		local obj = minetest.add_item(p, m_stack)

		if obj then
			obj:setvelocity({
				x = math.random(-2, 2),
				y = 7,
				z = math.random(-2, 2)
			})
		end
	end
end

**Comment by red-001** [comment]
@tenplus1 That's really something mods should handle using the on_blast callback. See #806.

**Comment by tenplus1** [comment]
red-001: the above code was for default chests in game either using tnt mod or edited for on_blast itself.

**Comment by red-001** [comment]
@tenplus1 I would agree minetest game should use on_blast more.

**Comment by paramat** [comment]
A dev has said there were reasons the voxelmanip wasn't used for removing nodes, you need to research why, try looking through the files history at previous commits. This PR cannot be merged until that research is done.

**Comment by red-001** [comment]
@paramat  I couldn't find any mention of the reasons in the file history or in the comments on #277  but like you said @ShadowNinja  said there where reasons for not using `voxelmanip`. ([link to IRC logs](http://irc.minetest.ru/minetest-dev/2016-01-31#i_4520541))  However at the time of the [last major rewrite of TNT](https://github.com/minetest/minetest_game/commit/0ca4520cc22fdea26e1d94800d112b957975e28d) the `on_blast` callback wasn't implanted. The `on_blast` callback has be implanted almost a year ago 8bc8dd6 `on_blast` can be used similarly  to `on_destruct` for stopping blocks form being destroyed by TNT. An `after_blast` callback could be added but mods could also remove the node themselves.

**Comment by paramat** [comment]
There are lots of changes coming for TNT it seems, and too close to release, sofar is busy with his stuff, so it seems the best thing is to delay TNT changes until after release, then do a thorough rewrite.

**Comment by sofar** [comment]
I'm counting 6 PRs with related changes. I'm rebasing all of them in a single PR, and making it a master WIP TNT PR that can be used to further improve TNT and comprehensively test all the changes collectively, which is rather cumbersome at this point.

**Comment by sofar** [comment]
Rebased and folded in to #862

**Pull request closed without merging**